crates/wasi-nn/src/backend/winml.rs

```diff
@@ -226,3 +205,180 @@ fn convert_i64(i: i64) -> Result<u32, BackendError> {
+            .as_ref()
+            .unwrap()
+            .cast::<TensorFloat16Bit>()
+            .unwrap();
+        let view = winml_tensor.GetAsVectorView().unwrap();
+        assert_eq!(view.into_iter().collect::<Vec<f32>>(), data);
+        // Convert back.
+        let t = to_tensor(inspectable.unwrap(), TensorKind::Float16);
+        assert!(t.as_ref().is_ok());
+        let t_ref = t.as_ref();
+        assert_eq!(t_ref.unwrap().data, buffer);
+        assert_eq!(t_ref.unwrap().dimensions, tensor.dimensions);
+        assert_eq!(t_ref.unwrap().ty, TensorType::Fp16);
+    }
+
+    #[test]
+    fn fp32() {
+        let data = vec![1.0_f32, 2.0, 3.0, 4.0, 5.0, 6.0];
+        let mut buffer = Vec::with_capacity(data.len() * size_of::<f32>());
+        for f in &data {
+            buffer.extend(f.to_ne_bytes());
+        }
+        let buffer_copy = buffer.clone();
+        let tensor = Tensor {
+            ty: TensorType::Fp32,
+            dimensions: vec![2, 3],
+            data: buffer_copy,
+        };
+        let shape = IVectorView::<i64>::try_from(vec![2i64, 3]).unwrap();
+        let inspectable = to_inspectable(&tensor, shape);
+        assert!(inspectable.is_ok());
+        let winml_tensor = inspectable.as_ref().unwrap().cast::<TensorFloat>().unwrap();
+        let view = winml_tensor.GetAsVectorView().unwrap();
+        assert_eq!(view.into_iter().collect::<Vec<f32>>(), data);
+        // Convert back.
+        let t = to_tensor(inspectable.unwrap(), TensorKind::Float);
+        assert!(t.as_ref().is_ok());
+        let t_ref = t.as_ref();
+        assert_eq!(t_ref.unwrap().data, buffer);
+        assert_eq!(t_ref.unwrap().dimensions, tensor.dimensions);
+        assert_eq!(t_ref.unwrap().ty, TensorType::Fp32);
+    }
+
+    #[test]
+    fn i64() {
+        let data = vec![6i64, 5, 4, 3, 2, 1];
+        let mut buffer = Vec::with_capacity(data.len() * size_of::<i64>());
+        for f in &data {
+            buffer.extend(f.to_ne_bytes());
+        }
+        let buffer_copy = buffer.clone();
+        let tensor = Tensor {
+            ty: TensorType::I64,
+            dimensions: vec![1, 6],
+            data: buffer_copy,
+        };
+        let shape = IVectorView::<i64>::try_from(vec![1i64, 6]).unwrap();
+        let inspectable = to_inspectable(&tensor, shape);
+        assert!(inspectable.is_ok());
+        let winml_tensor = inspectable
+            .as_ref()
+            .unwrap()
+            .cast::<TensorInt64Bit>()
+            .unwrap();
+        let view = winml_tensor.GetAsVectorView().unwrap();
+        assert_eq!(view.into_iter().collect::<Vec<i64>>(), data);
+        // Convert back.
+        let t = to_tensor(inspectable.unwrap(), TensorKind::Int64);
+        assert!(t.as_ref().is_ok());
+        let t_ref = t.as_ref();
+        assert_eq!(t_ref.unwrap().data, buffer);
+        assert_eq!(t_ref.unwrap().dimensions, tensor.dimensions);
+        assert_eq!(t_ref.unwrap().ty, TensorType::I64);
```

Comment:
It took me a second to understand that what we're trying to do is "roundtrip" a tensor through `to_inspectable` and back through `to_tensor`. To make this more clear, can you look at factoring out some of the duplication here?

Reply:
Implemented `PartialEq` for `Tensor`, so we don't need a lot of `assert_eq!` here.